File: leetcode/python/000239-sliding-window-maximum.py
from collections import deque


class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        output = []
        q = deque()
        l = r = 0

        while r < len(nums):
            while q and q[-1] < nums[r]:
                q.pop()

            q.append(nums[r])

            if (r + 1) >= k:
                output.append(q[0])

                if nums[l] == q[0]:
                    q.popleft()

                l += 1

            r += 1

        return output

File: leetcode/javascript/000239.sliding-window-maximum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 * Time complexity: O(n)
 * Space complexity: O(n)
 */
var maxSlidingWindow = function (nums, k) {
  const output = [];
  const deque = new Deque();
  let l = 0;
  let r = 0;

  while (r < nums.length) {
    while (!deque.isEmpty() && nums[r] > deque.back()) {
      deque.popBack();
    }

    deque.pushBack(nums[r]);

    if (r + 1 >= k) {
      output.push(deque.front());

      if (nums[l] === deque.front()) {
        deque.popFront();
      }

      l++;
    }

    r++;
  }

  return output;
};
